internal/db/definition_validation.go

```diff
@@ -36,23 +106,53 @@ var patchCollectionValidators = []func(
 	validateRootIDNotMutated,
 	validateSchemaVersionIDNotMutated,
 	validateCollectionNotRemoved,
+	validateSingleVersionActive,
+	validateSchemaNotAdded,
+	validateSchemaFieldNotDeleted,
+	validateFieldNotMutated,
+	validateFieldNotMoved,
 }
 
-var newCollectionValidators = []func(
-	client.CollectionDefinition,
-	map[string]client.CollectionDefinition,
-) error{
-	validateSecondaryFieldsPairUp,
+// globalValidators are run on create and update of records.
+var globalValidators = []definitionValidator{
+	validateCollectionNameUnique,
 	validateRelationPointsToValidKind,
+	validateSecondaryFieldsPairUp,
 	validateSingleSidePrimary,
+	validateCollectionDefinitionPolicyDesc,
+	validateSchemaNameNotEmpty,
+	validateRelationalFieldIDType,
+	validateSecondaryNotOnSchema,
+	validateTypeSupported,
+	validateTypeAndKindCompatible,
+	validateFieldNotDuplicated,
 }
 
+var updateValidators = append(
+	append([]definitionValidator{}, updateOnlyValidators...),
+	globalValidators...,
+)
+
+var createValidators = append(
+	append([]definitionValidator{}, createOnlyValidators...),
+	globalValidators...,
+)
+
 func (db *db) validateCollectionChanges(
-	oldColsByID map[uint32]client.CollectionDescription,
+	ctx context.Context,
+	oldCols []client.CollectionDescription,
 	newColsByID map[uint32]client.CollectionDescription,
 ) error {
-	for _, validators := range patchCollectionValidators {
-		err := validators(oldColsByID, newColsByID)
+	newCols := make([]client.CollectionDescription, 0, len(newColsByID))
+	for _, col := range newColsByID {
+		newCols = append(newCols, col)
+	}
+
+	newState := newDefinitionState(newCols, map[string]client.SchemaDescription{})
+	oldState := newDefinitionState(oldCols, map[string]client.SchemaDescription{})
+
+	for _, validator := range updateValidators {
+		err := validator(ctx, db, newState, oldState)
```

Comment:
question: would it be possible to return all validator errors? It would be helpful to know all of the problems with a schema / collection.

Reply:
Was this what you meant in https://github.com/sourcenetwork/defradb/issues/2568?

It should be fine and easy to do, although the resultant errors might be quite noisy (some errors will likely result in many other errors being returned).

Can we do this in another PR? Then we can test it properly (I want to avoid changing the tests as much as possible in this refactor PR)

Reply:
> Was this what you meant in https://github.com/sourcenetwork/defradb/issues/2568 (I wasn't sure if that ticket was for this, or some other validation somewhere)?

Yeah I was referencing that issue. 

> Can we do this in another PR/issue if we want it? Then we can test it properly (I want to avoid changing the tests as much as possible in this refactor PR)

Yeah that's fine with me. I think its purely a user experience issue and not relevant to this PR.